src/System.Text.Json/src/System/Text/Json/Serialization/JsonClassInfo.cs

```diff
@@ -311,9 +311,18 @@ public static Type GetElementType(Type propertyType)
                 elementType = propertyType.GetElementType();
                 if (elementType == null)
                 {
+                    Type[] args = propertyType.GetGenericArguments();
+
                     if (propertyType.IsGenericType)
                     {
-                        elementType = propertyType.GetGenericArguments()[0];
+                        if (typeof(IDictionary).IsAssignableFrom(propertyType))
+                        {
+                            elementType = args[1];
```

Comment:
Is this guaranteed to not throw IndexOutOfRangeException? If so, add a Debug.Assert.